### ISSUE TITLE
Duplicate messages on go-ssb log

### ISSUE THREAD

**Issue by mplorentz** [planetary-social/planetary-ios#957]
I am documenting here an issue that occurred to me last week. I think we will close it as `wontfix` because it's in go-ssb. But we should be aware of it, especially when migrating to scuttlego (https://github.com/planetary-social/scuttlego/issues/90).

After resyncing my feed a couple weeks ago it appears that go-ssb wrote some of my published messages twice. The duplicate messages had the same key, value, hash, signature, but different sequence numbers. I happened on my own feed three times. 

This bug caused several problems in the app. First, it forked my feed and prevented any messages I published from replicating with other clients. They wouldn't replicate because the signature of the duplicate messages are invalid. Fortunately since no one replicated the bad fork I was able to recover by deleting and resyncing my feed. I just lost the messages I published after the duplicates happened. Second, it caused the "View Source" button on messages to show the source for the wrong message.

Here is the raw debugger output for the three duplicate messages:

## Sequence number 631
```
(lldb) po String(cString: ssbGetRawMessage(feedRef, 631))
▿ Optional<String>
  - some : "{\n  \"previous\": \"%QfCA9/mMHmD8FPooL7z6LF+S8rqzWPzI9WOfTJEs+AA=.sha256\",\n  \"author\": \"@4wXR/KiJrkz9D2LPXpZl5XOLw+gYCoJW6p6rwFlI5yA=.ed25519\",\n  \"sequence\": 631,\n  \"timestamp\": 1665235310480,\n  \"hash\": \"sha256\",\n  \"content\": {\n    \"type\": \"post\",\n    \"text\": \"Just found https://cal.com which looks like a free, open source, self-hostable version of Calendly. Has anyone else used this? Their hosted version is completely free for individuals too, which is better than Calendly where you only get 1 event.\"\n  },\n  \"signature\": \"FjP9awdoQmBqtSTbGMVpjC2SnjR+fRiMiLefr1E2If/VuwUhYc72eBIKfZWR6t1rklMD+u2Q2bHGvpaeOc1XCQ==.sig.ed25519\"\n}"

(lldb) po String(cString: ssbGetRawMessage(feedRef, 632))
▿ Optional<String>
  - some : "{\"previous\":\"%QfCA9/mMHmD8FPooL7z6LF+S8rqzWPzI9WOfTJEs+AA=.sha256\",\"author\":\"@4wXR/KiJrkz9D2LPXpZl5XOLw+gYCoJW6p6rwFlI5yA=.ed25519\",\"sequence\":631,\"timestamp\":1665235310480,\"hash\":\"sha256\",\"content\":{\"type\":\"post\",\"text\":\"Just found https://cal.com which looks like a free, open source, self-hostable version of Calendly. Has anyone else used this? Their hosted version is completely free for individuals too, which is better than Calendly where you only get 1 event.\"},\"signature\":\"FjP9awdoQmBqtSTbGMVpjC2SnjR+fRiMiLefr1E2If/VuwUhYc72eBIKfZWR6t1rklMD+u2Q2bHGvpaeOc1XCQ==.sig.ed25519\"}"
```

## Sequence number 632
```
(lldb) po String(cString: ssbGetRawMessage(feedRef, 633))
▿ Optional<String>
  - some : "{\n  \"previous\": \"%0Wr5Deyf5+6Zp52365KR2PdladXb0zc/9lVkiR+gLJo=.sha256\",\n  \"author\": \"@4wXR/KiJrkz9D2LPXpZl5XOLw+gYCoJW6p6rwFlI5yA=.ed25519\",\n  \"sequence\": 632,\n  \"timestamp\": 1665281861817,\n  \"hash\": \"sha256\",\n  \"content\": {\n    \"vote\": {\n      \"link\": \"%yeuggl5+/CXij/UWlOitmy8CZYZJiK6sbsaI8w5RU9M=.sha256\",\n      \"value\": 1,\n      \"expression\": \"💜\"\n    },\n    \"root\": \"%yeuggl5+/CXij/UWlOitmy8CZYZJiK6sbsaI8w5RU9M=.sha256\",\n    \"branch\": [\n      \"%yeuggl5+/CXij/UWlOitmy8CZYZJiK6sbsaI8w5RU9M=.sha256\"\n    ],\n    \"type\": \"vote\"\n  },\n  \"signature\": \"A8aQQEXH4UyjkUdnntzKCvKj5fuQo4bvSN6KsTovpuq0WnOwYwwIxihmGfOm3N0WoKGD4ImqRwUVbkUDXt0rCg==.sig.ed25519\"\n}"

(lldb) po String(cString: ssbGetRawMessage(feedRef, 634))
▿ Optional<String>
  - some : "{\"previous\":\"%0Wr5Deyf5+6Zp52365KR2PdladXb0zc/9lVkiR+gLJo=.sha256\",\"author\":\"@4wXR/KiJrkz9D2LPXpZl5XOLw+gYCoJW6p6rwFlI5yA=.ed25519\",\"sequence\":632,\"timestamp\":1665281861817,\"hash\":\"sha256\",\"content\":{\"vote\":{\"link\":\"%yeuggl5+/CXij/UWlOitmy8CZYZJiK6sbsaI8w5RU9M=.sha256\",\"value\":1,\"expression\":\"💜\"},\"root\":\"%yeuggl5+/CXij/UWlOitmy8CZYZJiK6sbsaI8w5RU9M=.sha256\",\"branch\":[\"%yeuggl5+/CXij/UWlOitmy8CZYZJiK6sbsaI8w5RU9M=.sha256\"],\"type\":\"vote\"},\"signature\":\"A8aQQEXH4UyjkUdnntzKCvKj5fuQo4bvSN6KsTovpuq0WnOwYwwIxihmGfOm3N0WoKGD4ImqRwUVbkUDXt0rCg==.sig.ed25519\"}"
```

## Sequence number 652
```
(lldb) po String(cString: ssbGetRawMessage(feedRef, 654))
▿ Optional<String>
  - some : "{\n  \"previous\": \"%zXDi5N2jwaym49bVHCXCNTEM0HDU0PCJVGjkIzudjqc=.sha256\",\n  \"author\": \"@4wXR/KiJrkz9D2LPXpZl5XOLw+gYCoJW6p6rwFlI5yA=.ed25519\",\n  \"sequence\": 652,\n  \"timestamp\": 1665755529432,\n  \"hash\": \"sha256\",\n  \"content\": {\n    \"type\": \"post\",\n    \"root\": \"%13v6i3s98mTuk9y70f0XNfh9HYP8wNdaF1/3EXv9Dhg=.sha256\",\n    \"branch\": [\n      \"%13v6i3s98mTuk9y70f0XNfh9HYP8wNdaF1/3EXv9Dhg=.sha256\"\n    ],\n    \"text\": \"Lovely photo. Welcome!\"\n  },\n  \"signature\": \"HoZ9L1LJg7uilomsaP4p/RDaOeFCVKQegSybywUj3PRKOOOjEhsp+e0n5iqo2aJlQEDyQ/b4GdZS551le9vsCw==.sig.ed25519\"\n}"

(lldb) po String(cString: ssbGetRawMessage(feedRef, 655))
▿ Optional<String>
  - some : "{\"previous\":\"%zXDi5N2jwaym49bVHCXCNTEM0HDU0PCJVGjkIzudjqc=.sha256\",\"author\":\"@4wXR/KiJrkz9D2LPXpZl5XOLw+gYCoJW6p6rwFlI5yA=.ed25519\",\"sequence\":652,\"timestamp\":1665755529432,\"hash\":\"sha256\",\"content\":{\"type\":\"post\",\"root\":\"%13v6i3s98mTuk9y70f0XNfh9HYP8wNdaF1/3EXv9Dhg=.sha256\",\"branch\":[\"%13v6i3s98mTuk9y70f0XNfh9HYP8wNdaF1/3EXv9Dhg=.sha256\"],\"text\":\"Lovely photo. Welcome!\"},\"signature\":\"HoZ9L1LJg7uilomsaP4p/RDaOeFCVKQegSybywUj3PRKOOOjEhsp+e0n5iqo2aJlQEDyQ/b4GdZS551le9vsCw==.sig.ed25519\"}"
```

I was able to fix my profile by deleting and reinstalling Planetary. I also exported and saved my entire db in case we want to look into this more in the future.

**Comment by danlatorre** [comment]
Need to clarify scope & chat w/ Matt

**Comment by boreq** [comment]
@mplorentz 

I am pretty worried about this as we decided to not verify message signatures when importing them for performance reasons. Instead we simply trust that the messages are correct.

**Comment by boreq** [comment]
No, go-ssb should verify the signatures correctly. This is why for performance reasons we are just assuming that messages are correctly signed when importing them and skip the signature verification step.

**Comment by mplorentz** [comment]
@boreq ah I understand now, thanks.

**Comment by mplorentz** [comment]
For posterity: we closed this because we are about to migrate user data from go-ssb to scuttlego which won't have the bug that creates these duplicate messages. When scuttlego encounters a feed with duplicate messages it will copy the message only once. The scuttlego receive log however will end up with two entries pointing to the same message. In cases where scuttlego needs to get a single receive log sequence number for a duplicated message id it will return the lower sequence number. This should allow scuttlego to gracefully ignore the duplicate messages, i.e. when gossiping with other peers.  [Slack context](https://planetary-app.slack.com/archives/CE6RE7ND6/p1669996210844639)
